src/overmap.cpp

```diff
@@ -1329,84 +1334,182 @@ class joins_tracker
                             om_direction::type rot ) {
+            consistency_check();
             for( const std::pair<const cube_direction, std::string> &p : ter.joins ) {
                 cube_direction dir = p.first + rot;
                 const std::string &join = p.second;
 
                 pos_dir this_side{ pos, dir };
                 pos_dir other_side = this_side.opposite();
                 if( !other_side.inbounds() ) {
-                    orphaned.push_back( { this_side, join, 0 } );
+                    debugmsg( "out of bounds join" );
                     continue;
                 }
 
-                if( resolved_position_index.count( other_side ) ) {
+                if( resolved.count( other_side ) ) {
                     erase_unresolved( this_side );
-                } else if( postponed_points.count( pos ) ) {
-                    postponed.push_back( { other_side, join, 0 } );
                 } else {
+                    // If there were postponed joins pointing into this point,
+                    // so we need to un-postpone them because it might now be
+                    // possible to satisfy them.
+                    restore_postponed_at( other_side.p );
                     add_unresolved( other_side, join );
                 }
-                add_resolved( this_side, join );
+                resolved.add( *this, this_side, join );
             }
+            consistency_check();
         }
 
         std::pair<pos_dir, placement_constraints> pick_top_priority() const {
             cata_assert( any_unresolved() );
-            auto priority_it = std::find_if(
-                                   unresolved_priority_index.begin(), unresolved_priority_index.end(),
+            auto priority_it =
+                std::find_if( unresolved_priority_index.begin(), unresolved_priority_index.end(),
             []( const cata::flat_set<iterator, compare_iterators> &its ) {
                 return !its.empty();
             } );
             cata_assert( priority_it != unresolved_priority_index.end() );
             auto it = random_entry( *priority_it );
             const tripoint_om_omt &pos = it->where.p;
+            cata_assert( !postponed.any_at( pos ) );
             std::pair<pos_dir, placement_constraints> result( it->where, {} );
-            for( cube_direction dir : all_enum_values<cube_direction>() ) {
-                pos_dir key{ pos, dir };
-                auto pos_it = unresolved_position_index.find( key );
-                if( pos_it != unresolved_position_index.end() ) {
-                    result.second.joins.emplace_back( dir, pos_it->second->join_id );
-                }
+            for( iterator it : unresolved.all_at( pos ) ) {
+                result.second.joins.emplace_back( it->where.dir, it->join_id );
             }
             return result;
         }
         void postpone( const tripoint_om_omt &pos ) {
```

Comment:
Should there be something to make sure (what would normally be?) postponements are handled properly during the last phase?

Reply:
Are you thinking something like "consider it an error to postpone anything in the last phase"?  That will nearly always lead to an unresolved join error at the end, so I don't think we need extra checking on top of that.